src/types.md

```diff
@@ -236,3 +236,311 @@ fn main() {
+too much special about your type.
+
+## Should I have public fields or accessor methods?
+
+The trade-offs are similar to C++ except that Rust's pattern-matching makes it
+very convenient to match on fields, so within a realm of code that you own you
+may bias towards having more public fields than you're used to. As with C++,
+this can give you a future compatibility burden.
+
+## When should I use a newtype wrapper?
+
+The [newtype wrapper pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html)
+uses Rust's type systems to enforce extra behavior without necessarily changing
+the underlying representation.
+
+```rust
+# fn get_rocket_length() -> Inches { Inches(7) }
+struct Inches(u32);
+struct Centimeters(u32);
+
+fn build_mars_orbiter() {
+  let rocket_length: Inches = get_rocket_length();
+  // mate_to_orbiter(rocket_length); // does not compile because this takes cm
+}
+```
+
+Other examples that have been used:
+* An IP address which is guaranteed not to be localhost;
+* Non-zero numbers;
+* IDs which are guaranteed to be unique
+
+Such new types typically need a lot of boilerplate, especially to implement
+the traits which users of your type would expect to find. On the other hand,
+they allow you to use Rust's type system to statically prevent logic bugs.
+
+A heuristic: if there are some invariants you'd be checking for at runtime,
+see if you can use a newtype wrapper to do it statically instead. Although it
+may be more code to start with, you'll save the effort of finding and fixing
+logic bugs later.
+
+## How else can I use Rust's type system to avoid high-level logic bugs?
+
+Yes.
+
+*Zero sized types* ("ZSTs"). Often used as capability tokens - you can statically
+prove that code exclusively has the right to do something. For example:
+
+```rust
+struct PermissionToUseCamera;
+
+fn prompt_user_for_permission() -> Option<PermissionToUseCamera> {
+  // ...
+#  None
+}
+
+// Statically proven that no code calls this unless it's attempted to
+// get permission and dealt with the consequences of rejection.
+fn take_photo(permission: &PermissionToUseCamera) {
+  // ...
+}
+
+//fn main() { }
+```
+
+ZSTs can also be used to demonstrate _exclusive_ access to some resource.
+
+```rust
+struct RobotArmAccessToken;
+
+fn move_arm(token: &mut RobotArmAccessToken, x: u32, y: u32, z: u32) {
+  // ...
+}
+
+fn attach_car_door(token: &mut RobotArmAccessToken) {
+  move_arm(token, 3, 4, 6);
+  move_arm(token, 5, 3, 6);
+}
+
+fn install_windscreen(token: &mut RobotArmAccessToken) {
+  move_arm(token, 7, 8, 2);
+  move_arm(token, 1, 2, 3);
+}
+
+fn main() {
+  let mut token = RobotArmAccessToken; // ensure only one exists
+  attach_car_door(&mut token);
+  install_windscreen(&mut token);
+}
+```
+
+(the type system would prevent these operations happening in parallel).
+
+*Marker traits*. Indicate that a type meets certain invariants, so subsequent
```

Comment:
Since this answer is getting long, maybe we can use ### to make sub-sections that show up in the TOC, are easy to find by skimming, et c.?